sdk/eventhub/Microsoft.Azure.WebJobs.Extensions.EventHubs/src/Listeners/EventHubListener.PartitionProcessor.cs

```diff
@@ -227,19 +239,24 @@ private async Task TriggerExecute(EventData[] events, EventProcessorHostPartitio
                 };
 
                 await _executor.TryExecuteAsync(input, cancellationToken).ConfigureAwait(false);
+                _firstFunctionExecute = false;
             }
 
-            private async Task MonitorCachedEvents(CancellationTokenSource backgroundCancellationTokenSource)
+            private async Task MonitorCachedEvents(DateTimeOffset? lastCheckpointTime, CancellationTokenSource backgroundCancellationTokenSource)
             {
+                Debug.Assert(_minimumBatchesEnabled, "Monitoring of cached events should only take place when minimum batches are enabled.");
+
                 var acquiredSemaphore = false;
+                var adjustedMaxWaitTime = GetAdjustedMaxWaitTime(_maxWaitTime, lastCheckpointTime, _firstFunctionExecute);
+
                 try
                 {
                     _currentCycle = ValueStopwatch.StartNew();
```

Comment:
Just a thought but could it be helpful to start the timer at the time since the last checkpoint if it's less than the max wait time? If we don't, we may still run into the case where we wait .9999\*maxwaittime and then migrate and since we're only checking time since checkpoint on the first invocation, we could still end up waiting 1.999*maxwaittime? Unless I'm thinking about that wrong

Reply:
Not sure that I'm following your scenario.   To make it concrete, let's say that we have a wait time of 60 seconds and we read a checkpoint from 30 seconds ago.  Since checkpointing takes place after a batch is processed, we can infer that a function was invoked 30 seconds ago.  If we wanted to adjust to honor the wait time, we'd adjust the wait time to be 30 seconds.   That's what you're describing, right?

I actually had logic in for that at one point, but I ended up removing it to reduce complexity.   I've got the wait time documented to indicate that you may see 2 * wait time as when ownership changes.  Think it's important and that we should add that back in?

Reply:
Yeah that's what I was describing! And ok that makes sense. Having a maximum of 2maxwaittime for ownership changes sounds good to me since hopefully that perfectly timed migration won't be a very common issue.

Reply:
I added this back in for discussion.   Think we should keep it or pull it back out?

Reply:
I think I like it more this way personally, since it seems like checkpointfrequency=1 is the most common case, so preventing the potential 2maxwaittime wait seems like a positive feature.

Reply:
Alrighty.  We'll keep it then.